projects/rocprofiler-sdk/source/lib/output/generateRocpd.cpp

```diff
@@ -1441,6 +1442,273 @@ write_rocpd(
+                        ar(cereal::make_nvp("src_agent_id", agent_node_id(record.src_agent)));
+                        ar(cereal::make_nvp("dst_agent_id", agent_node_id(record.dst_agent)));
+                        ar(cereal::make_nvp("prefetch_agent_id",
+                                            agent_node_id(record.prefetch_agent)));
+                        ar(cereal::make_nvp("preferred_agent_id",
+                                            agent_node_id(record.preferred_agent)));
+                        ar(cereal::make_nvp("error_code", record.error_code));
+                    });
+                }
+                else if(std::holds_alternative<rocprofiler_buffer_tracing_kfd_page_fault_record_t>(
+                            itr.record))
+                {
+                    const auto& record =
+                        std::get<rocprofiler_buffer_tracing_kfd_page_fault_record_t>(itr.record);
+
+                    data.kind  = record.kind;
+                    data.name  = tool_metadata.buffer_names.at(data.kind, record.operation);
+                    data.tid   = record.pid;  // KFD attributes all events to the lead thread
+                    data.start = record.start_timestamp;
+                    data.end   = record.end_timestamp;
+                    data.value = record.address.value;
+
+                    data.json_data = get_json_string([&agent_node_id, &record](auto& ar) {
+                        ar(cereal::make_nvp("address", record.address.value));
+                        ar(cereal::make_nvp("agent_id", agent_node_id(record.agent_id)));
+                    });
+                }
+                else if(std::holds_alternative<rocprofiler_buffer_tracing_kfd_queue_record_t>(
+                            itr.record))
+                {
+                    const auto& record =
+                        std::get<rocprofiler_buffer_tracing_kfd_queue_record_t>(itr.record);
+
+                    data.kind  = record.kind;
+                    data.name  = tool_metadata.buffer_names.at(data.kind, record.operation);
+                    data.tid   = record.pid;  // KFD attributes all events to the lead thread
+                    data.start = record.start_timestamp;
+                    data.end   = record.end_timestamp;
+
+                    data.json_data = get_json_string([&agent_node_id, &record](auto& ar) {
+                        ar(cereal::make_nvp("agent_id", agent_node_id(record.agent_id)));
+                    });
+                }
+                else
+                {
+                    ROCP_FATAL << "unknown variant in tool_buffer_tracing_kfd_record_t";
+                }
+
+                // insert thread info if it doesn't already exist
+                get_thread_id(data.tid);
+
+                // create an empty event, just to associate the following info tables
+                auto evt_id = create_event(conn, {});
```

Comment:
```suggestion
                // get KFD category and create event entry
                auto category = tool_metadata.buffer_names.at(data.kind);
                auto evt_id =
                    create_event(conn,
                                 {
                                     insert_value("category_id", string_entries.at(category)),
                                 });
```
It is not recommended to simply insert an empty event.  The failed rocpd summary test is looking for a category. 

Even with this change, the rocpd_validation test is failing.  You may need to make some changes to the OTF2 output (in source/lib/python/rocpd/otf2.cpp) or the test validation otf2reader.py.

Reply:
Do you have a recommendation for how to link `rocpd_region` and `rocpd_pmc_event` without using an empty event?

Reply:
My code snippet above which adds `category` at a minimum, is what I recommend.  

You connected `rocpd_region` and `rocpd_pmc_event` correctly with the `event_id`, we just need `rocpd_event` to have `category_id` populated too.

Reply:
Thanks -- somehow I didn't see your snippet

Reply:
Before I dive further into this, do you have an opinion about whether KFD tracing should be implied by either `--sys-trace` and/or `--runtime-trace` runtime flags?

I believe what's happening is these validation tests are using `--runtime-trace`, which turns on KFD tracing, and the tests are not yet updated to handle KFD data